theano/tensor/opt.py

```diff
@@ -2591,6 +2591,64 @@ def local_adv_sub1_adv_inc_sub1(node):
     return [T.cast(y, node.outputs[0].dtype)]
 
 
+@register_specialize
+@register_stabilize
+@register_canonicalize
+@gof.local_optimizer([IncSubtensor,
+                      AdvancedIncSubtensor,
+                      AdvancedIncSubtensor1])
+def local_useless_incsubtensor_alloc(node):
+    """
+    Replaces an [Advanced]IncSubtensor[1], whose increment is an `alloc` of
+    a variable that has either broadcastable or matching dimensions with the
+    target variable that shall be incremented, by one that skips the
+    intermediate `alloc`.
+    """
+    if isinstance(node.op, (IncSubtensor,
+                            AdvancedIncSubtensor,
+                            AdvancedIncSubtensor1)):
+        x = node.inputs[0]
+        y = node.inputs[1]
+        i = node.inputs[2:]
+
+        if y.owner is not None and isinstance(y.owner.op, T.Alloc):
+            # z is the input of the Alloc op, i.e. T.alloc(z, <shape>)
+            z = y.owner.inputs[0]
+
+            # A non-empty list of conditions means that `z` has some
+            # non-broadcastable dimensions that cannot be checked statically
+            # for equal size with the corresponding dimension of `x`. Thus, a
+            # runtime check is necessary.
+            try:
+                cond = [T.eq(x.shape[k], z.shape[k])
```

Comment:
I think I incorrectly assumed implicit extra dimensions to be the trailing dimensions, but they are in fact the leading dimensions. So this is a good point:
```
z_shape = (1, ) * (y.ndim - z.ndim) + z.shape
z_broad = (True, ) * (y.ndim - z.ndim) + z.broadcastable
```
However, I'm still confused why you believe `x` and `z` (or `y`) should be assumed to have the same shape. This is what you're asserting, aren't you?
```
... not node.fgraph.shape_feature.same_shape(x, y, dim_x=k, dim_y=k)
```
Let's make an example for `AdvancedIncSubtensor1`. Let `x` be some vector of size 10, `i` be an integer vector of size 3 with values in the range `0,...,x.shape[0]-1` and `y = alloc(z, 5)` be a vector of size 5 with a scalar `z`. Then, what you're trying to avoid is that when the optimization kicks in and detects the alloc of `z`, it would remove the alloc and, thus, hide the shape error which would have occurred because `i` and `y` have different sizes. Right? So, in this case, we want to make sure that `y.shape[0] == i.shape[0]`, but in the general case, it would be `x[i].shape[k] == y.shape[k]` if dimension `k` of `z` is broadcastable, and if dimension `k` of `z` is non-broadcastable we should check that `x[i].shape[k] == z_shape[k]` (this is where, I believe, your code that builds `z_shape` and `z_broad` is exactly right and important). I actually had that part about `z_shape` and `z_broad` wrong, so that's good, but I believe we *cannot* compare `y` (or `z`) with `x` but should be comparing with `x[i]`. Referring to the example above, let's assume `y = alloc(z, 3)` (now the dimensions are good), when you compare `x` with `y`, you assert that `10 == 3` (that's the wrong comparison), but when we compare `x[i]` and `y`, we assert `3 == 3` (that's correct). Do you see what I mean? But apparently we can't do `same_shape(x[i], y, k, k)` because `x[i]` may not have been part of the graph before the optimization was triggered and, as far as I understand shape inference, the shape information is not available then.